godiard/wikipedia-activity
# Port to Python3, GTK3, port mwlib to *-python alts

# Port to Python3 - Wikipedia Activity
This activity is in **BETA**  testing. All testers are welcome.
## Known Issues:
* Tables not working
* References not working (will be removed)

## Installation
The installation of wikipedia activity is easier for the end user as its only necessary for them to install the `.xo` file. 
The minimal steps to make this possible is given below. Wikipedia Activity uses compressed `.bz` dumps in xml to create articles from wikipedia. 

* go to dumps.wikipedia.org and download the latest dump in your preferred language
![image](https://user-images.githubusercontent.com/48695438/72207331-3e720a80-34a9-11ea-972a-e83f1292d0c2.png)

* For my test purposes, I am going to download `simplewiki dump progress on 20200101` as of 2020-01-11. Thats the latest complete dump.
![image](https://user-images.githubusercontent.com/48695438/72207397-d839b780-34a9-11ea-89b9-d0125956a8c2.png)

* Download the `*.bz2` link:
![image](https://user-images.githubusercontent.com/48695438/72207413-00291b00-34aa-11ea-8dd1-0314a69e746b.png)
> Alternatively, if you are low on data, you can use a smaller `bz2` for testing, consider using a smaller package. For example, the pure English Wikipedia gives entire wikipedia dump, in compressed state is `less than 17GB` and the smallest may be as small as `40-70MB`

* Lets get things into action: Clone the github repo
```
git clone https://github.com/srevinsaju/wikipedia-activity
git checkout python3-ss
```
Periodically pull new changes, as its getting updated quite frequently
* **Install the Languages**
Get the downloaded bz2 file directory
Change you directory to the language you wish to develop for. 
For example,  I am developing `en_simple`, so
```
cd en_simple
cp ~/Downloads/*.bz2 .
```
* Extract the bz2 file 
```
bzip2 -d <your bzip filename here>
```
You might need to install bzip2, if its not found. Google it
* If there are a trailing numbers after the `.xml` consider removing it for better presentation and to prevent errors while creating a bundle 
For example, change `ensimple....wiki.xml3116641313` to `ensimple....wiki.xml` 

* Now verify that you now have two files, `search.db` and `ensimple....*.xml` in your folder

*  Process the dump file: Execute within the language folder
```
../tools2/pages_parser.py en_simple   # replace en_simple with your language
```
> NOTE: The process described on the wiki is a bit complicated. The code has been adjusted to make work easier. This is done on https://github.com/godiard/wikipedia-activity/pull/1/commits/cf6fa127c26e0b187ee7cf8712c85745af98796a

It might take minutes (for 70 MB files) to Hours for (300 MB files) and weeks for 16.5 GB database files. It completely depends on your CPU speed.  At this moment, `pages_parser`, **crawls** through the files to create an index, links, etc to enhance faster access.
Quoting Gonzalo Odiard
> With the spanish file and a bit more than 2.3M pages, this process takes aprox 1:30 hours in my system.

* You can now confirm that `*.links` , `*.templates` and `redirects ` are found. If the above process crashes with any error, these files will be created, but they are likely to be empty. Re-running the the above code will not help, you will have to cleanup the unwanted empty files generated

* You can now add or remove certain files from the blacklist or favorite txt. 's its use is assumed.

* Perform the command
```
../tools2/make_selection.py
```
Alternatively, you can ignore balcklist and favourite by 
```
../tools2/make_selection.py --all
```
* Create the index
```
../tools2/create_index.py en_simple
```
* Test the Index
```
../tools2/test_index.py en_simple Sun
```
The above command checks in `en_simple` configuration and extracts the wikitext of `Sun`
* Generate Activity.info (new)
```
./setup.py gen_act <path to .xml file>
```
* Check the contents of `activity/activity.info`. Make necessary changes
* Create the Bundle. c.f :
```
./setup.py dist_xo en_simple/simplewiki-20200101-pages-articles-multistream.xml
```
## Dependencies
![server](https://user-images.githubusercontent.com/48695438/72204844-4a4fd380-348d-11ea-8f42-4e35d9f7ef9f.png)

## Checklist
* [x] Port to Python3
* [x]  Port to GTK3
* [x] Fix server.py 
* [x] Remove `search-toolbar` dependencies on `browse-activity : seach-toolbar` 
* [x] Fix Search bar
* [x] Remove mwlib and other files 
* [x] Add other python alternative 
* [ ] Fix new tab
* [ ] Fix home page
* [ ] make it more sweet
* [ ] Final Fixes
* [ ] Test on Sugar DE
* [ ] Test on Sugar SOAS
* [ ] Make an automated wiki-builder
* [ ] Improve upstream documentation


## Credits
@Hrishi1999 for testing this on a new system and helping me find a lot of bugs
@quozl for global info 

## Disclaimer
This activity has been one of the most hardest tasks ever taken up my me. See the time taken to see how long this PR has been worked upon
